Expose "bounce impact" and Storage Server "version catch-up rate" metrics (#4770)
apple/foundationdb: documentation/sphinx/source/release-notes/release-notes-700.rst

Status
------
* Added ``commit_batching_window_size`` to the proxy roles section of status to record statistics about commit batching window size on each proxy. `(PR #4735) <https://github.com/apple/foundationdb/pull/4735>`_

> * Added ``bounce_impact`` to the recovery_state section of status to report if the cluster is bounceable and if not, the reason for why it is not bounceable. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_

Reviewer: I'm trying to find a way to say that we're reporting that bouncing is clean or not rather than completely unbounceable. What do you think of this?

```suggestion
* Added ``cluster.bounce_impact`` section to status to report if there will be any extra effects when bouncing the cluster, and if so, the reason for those effects. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_
```

Author: I see. I assume you don't like "cleanly bounceable". 

I am fine with your suggestion.

I admit I don't have a proper understanding of "bouncing" a FoundationDB cluster: maybe you can provide some details about it?

Reviewer: I don't necessarily dislike "cleanly bounceable", but it is probably something that requires a little explanation to understand from a user perspective. A documentation section about bouncing in general would probably be pretty useful, maybe it's something we should try to tackle the next time we do a push on docs.

As to some details about bouncing a cluster (restarting all or some of its processes), basically there are a few things to know in general. One is that if you bounce any of the transaction subsystem (logs, resolvers, proxies, master, cluster controller, etc.), then the cluster is going to go through a recovery. During that time (hopefully single digit seconds), the cluster will be unable to start new transactions or write new data. This is true regardless of when you bounce. 

After the initial recovery, there is some time spent consuming any old generations of logs so that you can get rid of them. If you don't complete this, then those old generations will be present on your next recovery, and the current generation from the current recovery will be added to them. This is why you want to avoid doing too many bounces before you reach full recovery -- if you do you are accumulating these old generations and there is a cost to that.

If you bounce storage servers the cluster can continue operating normally, and the storage server will rejoin after a short recovery period for itself. As long as that bounce is fast, it will be able to get back up to date with the most recent mutations pretty quickly. There are also some other processes (I think data distributor and ratekeeper) that can be bounced without causing a recovery.

When FDB is run with fdbmonitor, it can be configured to let you stage changes to the conf that won't apply until you manually bounce the process. If you have any staged changes in this mode, bouncing will pick them up.

There's a similar situation with upgrades depending on how you run FDB, such that you can stage a new binary for upgrade that won't be run until after you bounce FoundationDB. Again, in this case bouncing will cause you to start running a new version.

Author: Thanks for the explanation. A related question: what is/are the API(s)  for doing the bounce?

And, I update the release notes section.

Reviewer: We provide a `kill` command that can be accessed through `fdbcli`. Using this, you can issue a fairly simultaneous termination request to every process in the cluster. Assuming you are using `fdbmonitor` (or some equivalent process monitor), the processes can then be restarted very quickly.

You can also, of course, terminate the fdbserver processes through any normal mechanism provided by the OS.

Status
------
* Added ``commit_batching_window_size`` to the proxy roles section of status to record statistics about commit batching window size on each proxy. `(PR #4735) <https://github.com/apple/foundationdb/pull/4735>`_
* Added ``cluster.bounce_impact`` section to status to report if there will be any extra effects when bouncing the cluster, and if so, the reason for those effects. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_
* Added ``fetched_versions`` to the storage metrics section of status to report how fast a storage server is catching up in versions. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_

> * Added ``fetch_frequency`` to the storage metrics section of status to report the version fetching frequency of a storage server. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_

Reviewer: Minor edit to remove mention of versions.

```suggestion
* Added ``fetches_from_logs`` to the storage metrics section of status to report how frequently a storage server fetches updates from transaction logs. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_
```

Author: Done.

Thanks a lot for your feedback!

> * Added ``fetch_frequency`` to the storage metrics section of status to report how frequently a storage server fetches updates from transaction logs. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_

Reviewer: I might have given a bad suggestion here last time that didn't include a change to the metric name:

```suggestion
* Added ``fetches_from_logs`` to the storage metrics section of status to report how frequently a storage server fetches updates from transaction logs. `(PR #4770) <https://github.com/apple/foundationdb/pull/4770>`_
```

Author: My bad. I should have caught that. Addressed now.
